File: src/main/java/tech/omnidigit/gitexporter/helper/GitFilterTask.java
package tech.omnidigit.gitexporter.helper;

import com.intellij.openapi.progress.ProgressIndicator;
import com.intellij.openapi.progress.Task;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.vcs.VcsException;
import git4idea.GitCommit;
import git4idea.history.GitHistoryUtils;
import git4idea.repo.GitRepository;
import git4idea.repo.GitRepositoryManager;
import org.jetbrains.annotations.NotNull;
import tech.omnidigit.gitexporter.config.GitExportConfig;
import tech.omnidigit.gitexporter.ui.CommitPreviewDialog;

import java.time.Instant;
import java.time.LocalDateTime;
import java.time.ZoneId;
import java.time.format.DateTimeFormatter;
import java.time.format.DateTimeParseException;
import java.util.List;
import java.util.stream.Collectors;

/*
 * Copyright (c) 2025 Tony Ben
 */

public class GitFilterTask extends Task.Backgroundable {

    private final Project project;
    private final GitExportConfig config;

    private final CommitPreviewDialog previewDialog;

    public GitFilterTask(Project project, GitExportConfig config, CommitPreviewDialog previewDialog) {
        super(project, "Exporting Git Changes", true);
        this.project = project;
        this.config = config;
        this.previewDialog = previewDialog;
    }

    @Override
    public void run(@NotNull ProgressIndicator indicator) {
        // 实现导出逻辑
        GitRepository repo = findGitRepo();
        if (repo == null) {
            throw new IllegalStateException("No Git repository found in the project.");
        }

        previewDialog.addCommits(filterCommits(repo));
    }

    public GitRepository findGitRepo() {
        // 查找项目对应的 Git 仓库
        return GitRepositoryManager.getInstance(project).getRepositories().get(0);
    }

    public List<GitCommit> filterCommits(GitRepository repo) {
        // 根据配置筛选提交
        List<GitCommit> commits;
        try {
            commits = GitHistoryUtils.history(project, repo.getRoot());

            commits = commits.stream()
                    .filter(commit -> config.author() == null || config.author().isEmpty() ||
                            commit.getAuthor().getName().equals(config.author()))
                    .filter(commit -> config.message() == null || config.message().isEmpty() ||
                            commit.getSubject().contains(config.message()))
                    .filter(commit -> {
                        Instant commitTime = Instant.ofEpochSecond(commit.getAuthorTime());

                        // 处理开始日期
                        boolean startDateValid = true;
                        if (config.startDate() != null && !config.startDate().isEmpty()) {
                            try {
                                DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy/MM/dd HH:mm:ss");
                                LocalDateTime startDateTime = LocalDateTime.parse(config.startDate(), formatter);
                                Instant startDate = startDateTime.atZone(ZoneId.systemDefault()).toInstant();
                                startDateValid = !commitTime.isBefore(startDate);
                            } catch (DateTimeParseException e) {
                                e.printStackTrace();
                            }
                        }

                        // 处理结束日期
                        boolean endDateValid = true;
                        if (config.endDate() != null && !config.endDate().isEmpty()) {
                            try {
                                DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy/MM/dd HH:mm:ss");
                                LocalDateTime endDateTime = LocalDateTime.parse(config.endDate(), formatter);
                                Instant endDate = endDateTime.atZone(ZoneId.systemDefault()).toInstant();
                                endDateValid = !commitTime.isAfter(endDate);
                            } catch (DateTimeParseException e) {
                                e.printStackTrace();
                            }
                        }

                        return startDateValid && endDateValid;
                    })
                    .collect(Collectors.toList());

        } catch (VcsException e) {
            e.printStackTrace();
            throw new RuntimeException(e);
        }

        return commits;
    }

}

File: src/main/kotlin/tech/omnidigit/gitexporter/helper/GitFilterTask.kt
package tech.omnidigit.gitexporter.helper

import com.intellij.openapi.progress.ProgressIndicator
import com.intellij.openapi.progress.Task.Backgroundable
import com.intellij.openapi.project.Project
import com.intellij.openapi.vcs.VcsException
import git4idea.GitCommit
import git4idea.history.GitHistoryUtils
import git4idea.repo.GitRepository
import git4idea.repo.GitRepositoryManager
import tech.omnidigit.gitexporter.config.GitExportConfig
import tech.omnidigit.gitexporter.ui.CommitPreviewDialog
import java.time.Instant
import java.time.LocalDateTime
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import java.time.format.DateTimeParseException

/*
 * Copyright (c) 2025 Tony Ben
 */

class GitFilterTask(
    project: Project,
    private val config: GitExportConfig,
    private val previewDialog: CommitPreviewDialog
) : Backgroundable(project, "Git Exporter", true) {

    override fun run(indicator: ProgressIndicator) {
        val repo = findGitRepo() ?: throw IllegalStateException("No Git repository found in the project.")
        previewDialog.addCommits(filterCommits(repo))
    }

    private fun findGitRepo(): GitRepository? {
        return GitRepositoryManager.getInstance(project).repositories.firstOrNull()
    }

    private fun filterCommits(repo: GitRepository): List<GitCommit> {
        return try {
            GitHistoryUtils.history(project, repo.root)
                .filter { commit ->
                    matchesAuthor(commit) &&
                            matchesMessage(commit) &&
                            matchesDateRange(commit)
                }
        } catch (e: VcsException) {
            e.printStackTrace()
            throw RuntimeException(e)
        }
    }

    private fun matchesAuthor(commit: GitCommit): Boolean {
        return config.author.isEmpty() ||
                commit.author.name == config.author
    }

    private fun matchesMessage(commit: GitCommit): Boolean {
        return config.message.isEmpty() ||
                commit.subject.contains(config.message)
    }

    private fun matchesDateRange(commit: GitCommit): Boolean {
        val commitTime = Instant.ofEpochSecond(commit.authorTime)
        return isAfterStartDate(commitTime) && isBeforeEndDate(commitTime)
    }

    private fun isAfterStartDate(commitTime: Instant): Boolean {
        return config.startDate.takeIf { it.isNotEmpty() }?.let { dateStr ->
            parseDateTime(dateStr)?.let { startDate ->
                !commitTime.isBefore(startDate)
            } ?: true // If parsing fails, include the commit
        } ?: true // No start date filter
    }

    private fun isBeforeEndDate(commitTime: Instant): Boolean {
        return config.endDate.takeIf { it.isNotEmpty() }?.let { dateStr ->
            parseDateTime(dateStr)?.let { endDate ->
                !commitTime.isAfter(endDate)
            } ?: true // If parsing fails, include the commit
        } ?: true // No end date filter
    }

    private fun parseDateTime(dateTimeStr: String): Instant? {
        return try {
            DateTimeFormatter.ofPattern("yyyy/MM/dd HH:mm:ss")
                .parse(dateTimeStr)
                .let { LocalDateTime.from(it) }
                .atZone(ZoneId.systemDefault())
                .toInstant()
        } catch (e: DateTimeParseException) {
            e.printStackTrace()
            null
        }
    }
}